src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsSecretSource.java

```diff
@@ -0,0 +1,68 @@
+package io.jenkins.plugins.credentials.secretsmanager;
+
+import com.amazonaws.SdkClientException;
+import com.amazonaws.client.builder.AwsClientBuilder;
+import com.amazonaws.services.secretsmanager.AWSSecretsManager;
+import com.amazonaws.services.secretsmanager.AWSSecretsManagerClient;
+import com.amazonaws.services.secretsmanager.AWSSecretsManagerClientBuilder;
+import com.amazonaws.services.secretsmanager.model.AWSSecretsManagerException;
+import com.amazonaws.services.secretsmanager.model.GetSecretValueRequest;
+import com.amazonaws.services.secretsmanager.model.GetSecretValueResult;
+import com.amazonaws.services.secretsmanager.model.ResourceNotFoundException;
+import hudson.Extension;
+import io.jenkins.plugins.casc.SecretSource;
+import io.jenkins.plugins.credentials.secretsmanager.config.EndpointConfiguration;
+import io.jenkins.plugins.credentials.secretsmanager.config.PluginConfiguration;
+
+import java.io.IOException;
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@Extension
+public class AwsSecretSource extends SecretSource {
+
+    private static final Logger LOG = Logger.getLogger(AwsSecretSource.class.getName());
+
+    @Override
+    public Optional<String> reveal(String id) throws IOException {
+        final AWSSecretsManager client;
+        try {
+            client = createClient();
```

Comment:
Seems a little heavy to create a client on each call to reveal 😅

Consider moving the client to the `init` method.

Reply:
Please see this for reference: https://github.com/jenkinsci/hashicorp-vault-plugin/blob/master/src/main/java/com/datapipe/jenkins/vault/jcasc/secrets/VaultSecretSource.java

Reply:
I think I had done this to account for the fact that plugin config could be changed at any time through the Web UI. Does the init() handler get triggered when this happens?

(Of course if Jenkins is being managed with CasC then the user *shouldn't* be manually editing config through Web UI, but it's still technically *possible*.)

Reply:
NO... Just no!
`init` is called on each `reconfigure` so if someone changed the settings in between I be surprised.

init method is just that. initializing your secret source right before it is used.

https://github.com/jenkinsci/configuration-as-code-plugin/blob/f8dedcb8e5aed9890226c75608241dbfda444791/plugin/src/main/java/io/jenkins/plugins/casc/ConfigurationAsCode.java#L731

Reply:
I've tried shifting client initialisation to the init() method, but there appear to be problems when it interacts with the `@ConfiguredWithCode` rule in the tests.

It appears the rule is either not applying the config at the right time, or there could be a threading issue. The result is the plugin does not receive the CasC config, which in this case means it doesn't use the local AWS mock and the tests fail.

Reply:
@chriskilding that's why we use chained rules.
See 
https://github.com/jenkinsci/hashicorp-vault-plugin/blob/master/src/test/java/com/datapipe/jenkins/vault/jcasc/secrets/VaultSecretSourceTest.java#L54-L97

Reply:
Unfortunately it seems that the AWS SDK does not have an environment variable or Java system property for setting the service endpoint, so the `EnvVarsRule` will not work here.

Any other ideas?

Reply:
See how Azure Key vault does it, your assumption should as you cannot only be using plugin configuration you need to include environment variables and/or properties otherwise it will not support a JCasC setup for secrets.

https://github.com/jenkinsci/azure-keyvault-plugin/blob/ef3a5eb4413f7ecd47d6532735604c52fceaaef0/src/main/java/org/jenkinsci/plugins/azurekeyvaultplugin/AzureKeyVaultGlobalConfiguration.java#L78-L189

Reply:
Right, I've invented 2 env vars for endpoint configuration in the SecretSource:

- AWS_SERVICE_ENDPOINT
- AWS_SIGNING_REGION

These don't technically exist in the AWS SDK, but it would be quite helpful if they did! (Maybe I should propose them as changes on the SDK itself.)